### PR TITLE
ALTAPPS-680: Shared add progress into gamification toolbar

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[#ALTAPPS-680](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-680)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] Sentry Performance Monitoring of screen loading is added for new screens;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- `TrackWithProgress` added into `GamificattionToolbarFeature.State`
- StudyPlan & TopicCompleted subscription added to reload `TrackWithProgress`.
- `GamificattionToolbarScreen` added as a dependency for `GamificattionToolbarActionDispatcher` to use `fetchTrackProgressSentryTransaction` for isolated `TrackWithProgress` loading.

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarActionDispatcher.kt [hunk 1]
```diff
@@ -78,15 +113,59 @@ class GamificationToolbarActionDispatcher(
...
+            }
             is Action.LogAnalyticEvent ->
                 analyticInteractor.logEvent(action.analyticEvent)
             else -> {
                 // no op
             }
         }
     }
+
+    private suspend fun fetchTrackWithProgressThoughtStudyPlan(forceLoadFromRemote: Boolean): Result<TrackWithProgress?> =
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarActionDispatcher.kt:143]: I suppose here you mean `Thought` -> `Through`

File: shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarActionDispatcher.kt [hunk 2]
```diff
@@ -78,15 +113,59 @@ class GamificationToolbarActionDispatcher(
...
+            val studyPlan =
+                currentStudyPlanStateRepository.getState(forceLoadFromRemote).getOrThrow()
+            if (studyPlan.trackId != null) {
+                fetchTrackWithProgress(studyPlan.trackId, forceLoadFromRemote).getOrThrow()
+            } else {
+                null
+            }
+        }
+
+    private suspend fun fetchTrackWithProgress(trackId: Long, forceLoadFromRemote: Boolean): Result<TrackWithProgress?> =
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarActionDispatcher.kt:154]:
```suggestion
    private suspend fun fetchTrackWithProgress(
        trackId: Long,
        forceLoadFromRemote: Boolean
    ): Result<TrackWithProgress?> =
```

File: shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarActionDispatcher.kt [hunk 3]
```diff
@@ -78,15 +113,59 @@ class GamificationToolbarActionDispatcher(
...
+                    trackRepository.getTrack(trackId, forceLoadFromRemote)
+                }
+                val trackProgressDeferred = async {
+                    progressesRepository
+                        .getTrackProgress(trackId, forceLoadFromRemote)
+                }
+                TrackWithProgress(
+                    track = trackDeferred.await().getOrThrow(),
+                    trackProgress = trackProgressDeferred.await().getOrThrow()
+                        ?: return@runCatching null
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarActionDispatcher.kt:167]:
```suggestion
                    trackProgress = trackProgressDeferred.await().getOrThrow() ?: return@runCatching null
```

Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarActionDispatcher.kt:167]: It's more readable when in one line

Screenshot 2023-04-10 at 2 06 01 PM

File: shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarReducer.kt [hunk 4]
```diff
@@ -2,36 +2,56 @@ package org.hyperskill.app.gamification_toolbar.presentation
...
+                            state.copy(isRefreshing = true) to
+                                setOf(Action.FetchGamificationToolbarData(screen, true))
+                        }
                     is State.Error ->
-                        State.Loading to setOf(Action.FetchGamificationToolbarData(message.screen))
+                        State.Loading to setOf(Action.FetchGamificationToolbarData(screen, true))
                     else ->
                         null
                 }
+
```
[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarReducer.kt:46]: empty line

File: shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarReducer.kt [hunk 5]
```diff
@@ -2,36 +2,56 @@ package org.hyperskill.app.gamification_toolbar.presentation
...
+            is Message.FetchTrackWithProgressResult -> {
+                if (state is State.Content && message is Message.FetchTrackWithProgressResult.Success) {
+                    state.copy(trackWithProgress = message.trackWithProgress) to emptySet()
+                } else {
+                    null
+                }
+            }
+
```
[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarReducer.kt:54]: empty line

**Review by ivan-magda — CHANGES_REQUESTED** [review]
I think it's better not to merge this to `develop` because it adds new remote data loading without UI -> we don't want to call new remote data loading methods until the actual UI uses them.

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Profile/Views/Statistics/ProfileStatisticsItemView.swift [hunk 6]
```diff
@@ -38,8 +38,7 @@ struct ProfileStatisticsItemView: View {
         .cornerRadius(appearance.cornerRadius)
     }
 
-    @ViewBuilder
-    private var image: some View {
+    @ViewBuilder private var image: some View {
```
[the remark this answers is not in the document]
Comment (ivan-magda) [reply iosHyperskillApp/iosHyperskillApp/Sources/Modules/Profile/Views/Statistics/ProfileStatisticsItemView.swift:41]: This is maybe false positive, I'll look

Screenshot 2023-04-10 at 2 33 11 PM

**Comment by XanderZhu** [comment]
@ivan-magda 
> I think it's better not to merge this to develop because it adds new remote data loading without UI -> we don't want to call new remote data loading methods until the actual UI uses them.

Sure, we can create a Progresses feature branch to collect all the changes there.

File: shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarReducer.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/gamification_toolbar/presentation/GamificationToolbarReducer.kt:54]: Do you use custom linter settings?
`./gradlew ktlintCheck` doesn't show this issues

**Comment by XanderZhu** [comment]
@ivan-magda 
Swift lint warning @ViewBuilder issues
https://github.com/hyperskill/mobile-app/actions/runs/4659392411/jobs/8246224372?pr=430

**Review by ivan-magda — APPROVED** [review]

**Pull request merged**